### PR TITLE
Scripts/Naxxramas: Frogger event improvements

### PR DESCRIPTION
**Pull request by Jildor**
**Changes proposed:**

-  Prevent Living Poison accumulation
-  Now Living Poison cast Explode
-  Updates: https://github.com/TrinityCore/TrinityCore/issues/19072#issuecomment-667949712

~~Edit: There is still a problem when you are in the instance, but you are away from the npc spawn zone, because event still run but NPC doesn't despawn, I have no idea how to solve that properly, so I prefer not to touch that.~~

Edit2: I found a way, I hope can be valid ;)

**Target branch(es):** 3.3.5/master

- [X] 3.3.5
- [ ] master

**Issues addressed:** None

**Tests performed:** tested in-game

### CODE REVIEW

**Comment by Treeston** [comment]
if there is a trigger, just have the trigger spawn them in its own updateai

**Comment by offl** [comment]
You can't set unique waypoints to unique summoned creatures

**Comment by Treeston** [comment]
what i mean is take that entire event's logic with summoncreaturegroup and move it to the frogger trigger's AI instead

that way it only runs if that grid is loaded in the first place. problem solved, no grid searchers needed.

**Comment by Jildor** [comment]
thx, let me try

**Comment by Jildor** [comment]
updated

**Comment by offl** [comment]
Now we can discuss can we summon 3 mobs instantly without delays or not :D

**Comment by Jildor** [comment]
I prefer don't touch that cosmetic part xD
anyone can touch it up later

File: src/server/scripts/Northrend/Naxxramas/naxxramas.cpp [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+#include "ScriptedCreature.h"
+#include "TemporarySummon.h"
+#include "naxxramas.h"
+
+enum NaxxEvents
+{
+    EVENT_SUMMON_FROGGER_WAVE = 1
+};
+
+struct npc_frogger_trigger_naxx : public ScriptedAI
```
Comment (Carbenium) [inline src/server/scripts/Northrend/Naxxramas/naxxramas.cpp:32]: Is there a specific reason why this can't be in `instance_naxxramas.cpp`?

Comment (Jildor) [reply src/server/scripts/Northrend/Naxxramas/naxxramas.cpp:32]: I've seen other instance models and I think it should be separate, isn't it correct?

Comment (Carbenium) [reply src/server/scripts/Northrend/Naxxramas/naxxramas.cpp:32]: Hmm, indeed. That has slipped my attention.

Alright, carry on ;)

File: src/server/scripts/Northrend/Naxxramas/naxxramas.cpp [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+{
+    npc_frogger_trigger_naxx(Creature* creature) : ScriptedAI(creature), _instance(creature->GetInstanceScript()) { }
+
+    void Reset() override
+    {
+        _events.Reset();
+        _events.ScheduleEvent(EVENT_SUMMON_FROGGER_WAVE, 1s);
+        nextFroggerWave = 0;
```
Comment (jackpoz) [inline src/server/scripts/Northrend/Naxxramas/naxxramas.cpp:40]: please initialize this class member in the constructor too

**Review by jackpoz — CHANGES_REQUESTED** [review]

File: src/server/scripts/Northrend/Naxxramas/instance_naxxramas.cpp [hunk 3]
```diff
@@ -420,6 +417,9 @@ class instance_naxxramas : public InstanceMapScript
 
             void Update(uint32 diff) override
             {
+                if (!instance->HavePlayers())
+                    return;
+
```
Comment (Carbenium) [inline src/server/scripts/Northrend/Naxxramas/instance_naxxramas.cpp:422]: Is this needed with the trigger now being responsible for spawning the waves?

Comment (Jildor) [reply src/server/scripts/Northrend/Naxxramas/instance_naxxramas.cpp]: now it seems unnecessary

File: src/server/scripts/Northrend/Naxxramas/naxxramas.cpp [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+                default:
+                    break;
+            }
+        }
+    }
+
+private:
+    InstanceScript* _instance;
+    EventMap _events;
+    int8 _nextFroggerWave;
```
Comment (Treeston) [inline src/server/scripts/Northrend/Naxxramas/naxxramas.cpp:78]:
```suggestion
    uint8 _nextFroggerWave;
```

File: src/server/scripts/Northrend/Naxxramas/naxxramas.cpp [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+enum NaxxEvents
+{
+    EVENT_SUMMON_FROGGER_WAVE = 1
+};
+
+struct npc_frogger_trigger_naxx : public ScriptedAI
+{
+    npc_frogger_trigger_naxx(Creature* creature) : ScriptedAI(creature)
+    {
+        _instance = creature->GetInstanceScript();
```
Comment (Treeston) [inline src/server/scripts/Northrend/Naxxramas/naxxramas.cpp:35]:
```suggestion
```

File: src/server/scripts/Northrend/Naxxramas/naxxramas.cpp [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+#include "naxxramas.h"
+
+enum NaxxEvents
+{
+    EVENT_SUMMON_FROGGER_WAVE = 1
+};
+
+struct npc_frogger_trigger_naxx : public ScriptedAI
+{
+    npc_frogger_trigger_naxx(Creature* creature) : ScriptedAI(creature)
```
Comment (Treeston) [inline src/server/scripts/Northrend/Naxxramas/naxxramas.cpp:33]:
```suggestion
    npc_frogger_trigger_naxx(Creature* creature) : ScriptedAI(creature), _nextFroggerWave(0)
```

File: src/server/scripts/Northrend/Naxxramas/naxxramas.cpp [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    {
+        _instance = creature->GetInstanceScript();
+        Initialize();
+    }
+
+    void Initialize()
+    {
+        _nextFroggerWave = 0;
+    }
+
```
Comment (Treeston) [inline src/server/scripts/Northrend/Naxxramas/naxxramas.cpp:43]:
```suggestion
```

File: src/server/scripts/Northrend/Naxxramas/naxxramas.cpp [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+{
+    EVENT_SUMMON_FROGGER_WAVE = 1
+};
+
+struct npc_frogger_trigger_naxx : public ScriptedAI
+{
+    npc_frogger_trigger_naxx(Creature* creature) : ScriptedAI(creature)
+    {
+        _instance = creature->GetInstanceScript();
+        Initialize();
```
Comment (Treeston) [inline src/server/scripts/Northrend/Naxxramas/naxxramas.cpp:36]:
```suggestion
```

File: src/server/scripts/Northrend/Naxxramas/naxxramas.cpp [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    void Initialize()
+    {
+        _nextFroggerWave = 0;
+    }
+
+    void Reset() override
+    {
+        _events.Reset();
+        _events.ScheduleEvent(EVENT_SUMMON_FROGGER_WAVE, 1s);
+        Initialize();
```
Comment (Treeston) [inline src/server/scripts/Northrend/Naxxramas/naxxramas.cpp:48]:
```suggestion
```

File: src/server/scripts/Northrend/Naxxramas/naxxramas.cpp [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+                    break;
+                }
+                default:
+                    break;
+            }
+        }
+    }
+
+private:
+    InstanceScript* _instance;
```
Comment (Treeston) [inline src/server/scripts/Northrend/Naxxramas/naxxramas.cpp:76]:
```suggestion
```

**Comment by jackpoz** [comment]
Anything left to do in this PR ? Is it still working after all the changes ?

**Comment by Jildor** [comment]
is complete and tested

**Review by jackpoz — APPROVED** [review]

**Comment by jackpoz** [comment]
Thanks for the PR :)

**Pull request merged**